src/crypto/slow-hash.c

```diff
@@ -59,35 +60,104 @@ extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *exp
   } while(0)
 
 #define VARIANT1_2(p) \
-  do if (variant > 0) \
+  do if (variant == 1) \
   { \
     xor64(p, tweak1_2); \
   } while(0)
 
 #define VARIANT1_CHECK() \
   do if (length < 43) \
   { \
-    fprintf(stderr, "Cryptonight variants need at least 43 bytes of data"); \
+    fprintf(stderr, "Cryptonight variant 1 needs at least 43 bytes of data"); \
     _exit(1); \
   } while(0)
 
 #define NONCE_POINTER (((const uint8_t*)data)+35)
 
 #define VARIANT1_PORTABLE_INIT() \
   uint8_t tweak1_2[8]; \
-  do if (variant > 0) \
+  do if (variant == 1) \
   { \
     VARIANT1_CHECK(); \
     memcpy(&tweak1_2, &state.hs.b[192], sizeof(tweak1_2)); \
     xor64(tweak1_2, NONCE_POINTER); \
   } while(0)
 
 #define VARIANT1_INIT64() \
-  if (variant > 0) \
+  if (variant == 1) \
   { \
     VARIANT1_CHECK(); \
   } \
-  const uint64_t tweak1_2 = variant > 0 ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+  const uint64_t tweak1_2 = (variant == 1) ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+
+#define VARIANT2_INIT() \
+  uint32_t division_result[2]; \
+  uint32_t sqrt_result; \
+  do if (variant == 2) \
+  { \
+    *((uint64_t*)division_result) = 0; \
```

Comment:
No need for the aliasing violation, just do `division_result = {0, 0};`. Detecting multi-wide initialization is trivial for the compiler.